crates/daphne-server/src/router/mod.rs

```diff
@@ -324,18 +327,97 @@ where
             (task_id, resource)
         };
 
-        Ok(DapRequestExtractor(DapRequest {
+        let request = DapRequest {
             version,
             task_id,
             resource,
             payload: payload.to_vec(),
             media_type,
-            sender_auth: Some(sender_auth),
             taskprov,
-        }))
+        };
+
+        Ok(UnauthenticatedDapRequestExtractor(request))
+    }
+}
+
+/// An axum extractor capable of parsing a [`DapRequest`].
+///
+/// This extractor asserts that the request is authenticated.
+#[derive(Debug)]
+struct DapRequestExtractor(pub DapRequest);
+
+#[async_trait]
+impl<S, B> FromRequest<S, B> for DapRequestExtractor
+where
+    S: DaphneService + Send + Sync,
+    B: HttpBody + Send + 'static,
+    <B as HttpBody>::Data: Send,
+{
+    type Rejection = (StatusCode, String);
+
+    async fn from_request(req: Request<B>, state: &S) -> Result<Self, Self::Rejection> {
+        let bearer_token = extract_header_as_string(req.headers(), http_headers::DAP_AUTH_TOKEN)
+            .map(BearerToken::from);
+        let cf_tls_client_auth = extract_header_as_string(req.headers(), "X-Client-Cert-Verified");
+
+        let request = UnauthenticatedDapRequestExtractor::from_request(req, state)
+            .await?
+            .0;
+
+        let bearer_authed = if let Some((token, sender)) =
+            bearer_token.zip(request.media_type.map(|m| m.sender()))
+        {
+            state
+                .server_metrics()
+                .auth_method_inc(metrics::AuthMethod::BearerToken);
+            state
+                .check_bearer_token(
+                    &token,
+                    sender,
+                    request.task_id.unwrap(), // task ids are mandatory
+                    request.taskprov.is_some(),
+                )
+                .await
+                .map_err(|reason| {
+                    reason.either(
+                        |reason| (StatusCode::UNAUTHORIZED, reason),
+                        |_| (StatusCode::INTERNAL_SERVER_ERROR, "internal error".into()),
+                    )
+                })?;
+            true
+        } else {
+            false
+        };
+        let mtls_authed = if let Some(verification_result) = cf_tls_client_auth {
```

Comment:
Right now mutual TLS is only supported for the taskprov owner. We should therefore only use a client certificate to authorize access to the taskprov tasks.